Introduce a modern, JSX-like syntax for TwigComponents

Halleck45 opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes 
| Docs?         | yes 
| Issues        | 
| License       | MIT


Hey! 👋  

Following [this PR](https://github.com/symfony/ux/pull/2539), I have a small commit I'd love to discuss — I believe it goes hand-in-hand with the idea of introducing a **Toolkit**.

My proposal is about improving the **HTML syntax** used for TwigComponents, and **introducing a `JSX-like syntax`**

Currently, we write components like this:

```html
<twig:foo xxx=""></twig:foo>
```

My suggestion is to allow a more modern and readable syntax:

```html
<Foo xxx=""></Foo>
```

I strongly believe this new syntax would significantly improve readability.

Here’s a real-world example from one of my projects:

Before:

```html
<twig:PageTitle title="..." />

<twig:Button small  >
    <twig:ux:icon .../>
</twig:Button>

<twig:Section mdCols="2">
    <twig:BasicCard title="...">
    </twig:BasicCard>

    <twig:BasicCard title="...">
    </twig:BasicCard>
</twig:Section>
```

After:

```html
<PageTitle title="..." />

<Button small >
    <Ux:icon .../>
</Button>

<Section mdCols="2">
    <BasicCard title="...">
    </BasicCard>

    <BasicCard title="...">
    </BasicCard>
</Section>
```

To me, this looks cleaner, more readable, and more aligned with modern expectations (especially for developers familiar with React, etc.).

**💡 Backward compatibility note:**

This new syntax is entirely optional and fully backward compatible. The **JSX-like tags are only interpreted as Twig components if the tag name starts with an uppercase letter**. Tags starting with lowercase letters continue to behave as standard HTML or follow the current `<twig:...>` syntax rules.

The commit is super simple — which probably means I'm missing something 😅 — but before digging deeper, I’d love to get your thoughts on this direction.

Thanks!

Kocal commented on src/TwigComponent/src/Twig/TwigPreLexer.php:
```diff
@@ -35,10 +35,27 @@ public function __construct(int $startingLine = 1)
         }
 
+        if ($isJsx) {
+            $componentNames = array_map(fn ($match) => $match[1], $matches);
+            $componentNames = array_unique(array_filter($componentNames));
```
Instead, can we use a single `array_reduce` to loop over `$componentNames` only 1 time instead of 3 times?

Kocal commented on src/TwigComponent/src/Twig/TwigPreLexer.php:
```diff
@@ -35,10 +35,27 @@ public function __construct(int $startingLine = 1)
         }
 
+        if ($isJsx) {
+            $componentNames = array_map(fn ($match) => $match[1], $matches);
+            $componentNames = array_unique(array_filter($componentNames));
+
+            // To simplify things in the rest of the class, we replace the component name with twig:<componentName>
+            foreach ($componentNames as $componentName) {
```
Or maybe better, 1 loop instead of 4?  
```suggestion
            // To simplify things in the rest of the class, we replace the component name with twig:<componentName>
            foreach ($matches as $match) {
                if (!isset($componentName = $match[1])) {
                    continue;
                }
```

smnandre replied:
I think it would be even easier in the other way.... not sure we need the 2step PATH

Halleck45 replied:
For me, changing the code might actually make it slower. Your solution helps reduce iterations if all the nodes are unique.

But if a page contains the same node multiple times, it will trigger multiple `preg_replace` calls - which are much more expensive than a simple array iteration.

I think it's better to first reduce the array to a unique list, then perform the minimal number of `preg_replace` calls.

That said, happy to discuss it of course!

smnandre commented:
As i told you ... it's --personal stand point-- a _total_ no-go for me before 3.0.

Regarding the "new" syntax, am i missing something or is just the prefix ignored ? I mean, are there other features ? 

But I would **very much** use something like it   -- with a couple of maybe "control"  / safeguards but  --  yeah the DX **is** better without twig:)

Kocal reviewed (commented):
Cool :eyes:

smnandre commented:
> This new syntax is entirely optional and fully backward compatible.

No it is sadly not (at all).

When Twig HTML syntax was added, no one (we can bet on this) ths followinh thing in their templates:

<twig:A 

(Openin tag, followed by "twig"; follwed by colon, followed by letters)

Here it will have incidence on a lot of people, so yeah we cannot push it everywhere for everyone like that.


As you said, we can and should offer a flag/toggler/etc in the first time... But then we must not update already the documentation,  🤷‍♂️ .



So i'm still on the same mood: I know and agree it's cool §told you I already did it..)  But here it's not the part where the questions are. ALmost the oppostie, eve.  not a technical problem in usage, it's in term of impact, in clarity, in future rebustness and

Kheuval commented:
Hello, I like this idea a lot !

In Vue there is this [eslint rule](https://eslint.vuejs.org/rules/multi-word-component-names), I believe it would be nice to promote multi-words component names to improve readability even more and distinguish at a glance what is HTML and what is user-created.
This could take the form of some kind of warning triggered if a component is named using a single word.
In your example, `Foo` would become `AppFoo` for example.

What do you think ?

Kocal commented on src/TwigComponent/src/Twig/TwigPreLexer.php:
```diff
@@ -35,10 +35,27 @@ public function __construct(int $startingLine = 1)
+        // - jsx: <ComponentName> (with a capital letter)
+
+        $isClassical = str_contains($input, '<twig:');
+        $isJsx = preg_match_all('/<([A-Z][a-zA-Z0-9_:-]+)([^>]*)>/', $input, $matches, \PREG_SET_ORDER);
```
I don't think if we can really qualify it as JSX, the properties and spread syntaxes are different and here only the naming case is the same.
What about "$isShort" and "$isLong"?

alexander-schranz replied:
can it be both? Then `has` is better as `is`: https://3v4l.org/Wln8L

Halleck45 replied:
Thanks! To me, it doesn't seem relevant to use a `has` naming here. The goal is to detect which syntax is being used, not whether both are present.

If we ever introduce support for mixing syntaxes or handling them differently in the same input, then switching to a `hasShortSyntax` / `hasLongSyntax` style might make sense. But for now, a simple `$isShort` better reflects the current logic and avoids overcomplicating things prematurely.

alexander-schranz commented:
Just as here are some discussion about only add this to a new major. It could be possible to optin into this via something like:

```jinja
{% twig_component_mode 'experimental:short' %}
```

Inspired by the [trans_default_domain](https://symfony.com/doc/current/translation.html#using-twig-filters).

weaverryan commented:
I like the idea of making this "opt in" for now. That would allow people to start playing with this.
Personally I'd rather have a global config vs declaring at the top of the template

Halleck45 commented:
Hi! I’ve made the syntax disabled by default and added a global `short_syntax` option.

I also took the liberty of enhancing the `debug:twig-component` command while I was at it - in case we get any user feedback, I thought it would make it easier to help them debug the issue.

yguedidi commented on src/TwigComponent/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## 2.21.0
+
+-  Introduce an experimental JSX-like syntax for TwigComponents, making `<twig:` prefix optional
```
```suggestion
-  Introduce an experimental JSX-like syntax for TwigComponents, making `twig:` prefix optional
```

smnandre replied:
```suggestion
-  Add experimental support for prefix-less tags in HTML Syntax (`<FooBar />` instead of `<twig:FooBar />`)
```

yguedidi commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1719,6 +1719,36 @@ Pass the name of some component as an argument to print its details:
+
+An experimental short syntax is available for components, and has been introduced in 2.21.
+
+This mode allows you to omit the `<twig:` prefix and reference components directly by their name,
```
```suggestion
This mode allows you to omit the `twig:` prefix and reference components directly by their name,
```

yguedidi commented on src/TwigComponent/src/Command/TwigComponentDebugCommand.php:
```diff
@@ -85,6 +87,7 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
         $components = $this->findComponents();
         $this->displayComponentsTable($io, $components);
+        $this->displayDetailsAboutConfiguration($io);
```
shouldn't the configuration appear first, above components?

yguedidi replied:
btw, worth being a separate PR IMO (without the new setting of course), may get merged faster than this one

smnandre replied:
Yes it will be better if you can split this @Halleck45 thx!

yguedidi commented on src/TwigComponent/src/DependencyInjection/TwigComponentExtension.php:
```diff
@@ -217,6 +229,10 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultValue('%kernel.debug%')
                 ->end()
+                ->booleanNode('short_syntax')
+                    ->info('Enables the short syntax for Twig Components (the <twig: prefix is optional)')
```
```suggestion
                    ->info('Enables the short syntax for Twig Components (the `twig:` prefix is optional)')
```

yguedidi commented on src/TwigComponent/src/DependencyInjection/TwigComponentExtension.php:
```diff
@@ -129,18 +129,30 @@ static function (ChildDefinition $definition, AsTwigComponent $attribute) {
 
         $container->register('ux.twig_component.twig.lexer', ComponentLexer::class);
+        $container->getDefinition('ux.twig_component.twig.lexer')
+            ->addMethodCall('enableShortSyntax', [$config['short_syntax']]);
```
to me:
- either call the method `enableShortSyntax` but with no parameter, and have the `addMethodCall` wrapped by an `if` using the config value
- either call the method `setShortSyntax` and keep the parameter

I'd favor the first personally

yguedidi commented on src/TwigComponent/src/Twig/ComponentLexer.php:
```diff
@@ -26,9 +26,11 @@
  */
 class ComponentLexer extends Lexer
 {
+    private $withShortSyntax = false;
```
```suggestion
    private $allowingShortSyntax = false;
```
sounds better IMO

yguedidi commented on src/TwigComponent/src/Twig/TwigPreLexer.php:
```diff
@@ -28,17 +28,34 @@ class TwigPreLexer
-        if (!str_contains($input, '<twig:')) {
+        // tag may be:
+        // - long: <twig:componentName>
+        // - short (jsx like): <ComponentName> (with a capital letter)
```
```suggestion
        // - short (JSX-like): <ComponentName> (with a capital first letter)
```

yguedidi commented on src/TwigComponent/src/Twig/TwigPreLexer.php:
```diff
@@ -28,17 +28,34 @@ class TwigPreLexer
+            $componentNames = array_map(fn ($match) => $match[1], $matches);
+            $componentNames = array_unique(array_filter($componentNames));
+
+            // To simplify things in the rest of the class, we replace the component name with twig:<componentName>
```
```suggestion
            // To simplify things in the rest of the class, we replace the component name with <twig:componentName>
```

yguedidi reviewed (changes requested):
small remarks from me :)

smnandre commented:
> I'd rather have a global config vs declaring at the top of the template

So if there is a problem with only one file from a distant bundle resource, it would prevent to use the short syntax anywhere ?

smnandre commented on src/TwigComponent/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## 2.21.0
```
```suggestion
## 2.21.0
```

`2.24` i think ?  (feel free to create a new section here)

smnandre commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1719,6 +1719,36 @@ Pass the name of some component as an argument to print its details:
+Short Syntax
+------------
+
+An experimental short syntax is available for components, and has been introduced in 2.21.
```
Let's use a "versionadded" for this

smnandre commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1719,6 +1719,36 @@ Pass the name of some component as an argument to print its details:
     |                                                   | int $min = 10                     |
     +---------------------------------------------------+-----------------------------------+
 
+Short Syntax
```
This is **not** a new syntax.

We really should be precise on this :) Having _two_ syntaxes is already something that did create its fair share of frustrations, documetation problems, tests, etc etc ..  let's not frighten or confuse anyone with no reason here :)

smnandre commented on src/TwigComponent/doc/index.rst:
```diff
@@ -1719,6 +1719,36 @@ Pass the name of some component as an argument to print its details:
+
+    # config/packages/twig_component.yaml
+    twig_component:
+        short_syntax: true
```
`prefixless_tags` or `short_tags`  ?

smnandre commented on src/TwigComponent/src/Command/TwigComponentDebugCommand.php:
```diff
@@ -356,4 +359,17 @@ private function getAnonymousComponentProperties(ComponentMetadata $metadata): a
         return $properties;
     }
+
+    private function displayDetailsAboutConfiguration(SymfonyStyle $io): void
+    {
+        $io->section('Configuration of TwigComponent');
+        $io->table(
+            ['Configuration', 'Current value'],
+            [
+                ['anonymous_template_directory', $this->anonymousDirectory],
+                ['short_syntax', $this->withShortSyntax ? 'enabled' : 'disabled'],
+                ['profiler',  $this->withProfiler ? 'enabled' : 'disabled'],
+            ]
+        );
+    }
```
Not 100% sure this will bring more value than the `debug:config` command (a bit more complete / with examples, etc...)

smnandre commented on src/TwigComponent/src/DependencyInjection/TwigComponentExtension.php:
```diff
@@ -129,18 +129,30 @@ static function (ChildDefinition $definition, AsTwigComponent $attribute) {
+        // So we add setters for our required options
+        // This should be improved in the future : currently, some parameters of the ComponentLexer are not settable
+        $container->getDefinition('ux.twig_component.twig.environment_configurator')
+            ->addMethodCall('enabledShortSyntax', [$config['short_syntax']]);
```
Configurator should not have `itself` a setter, let's inject value IF we need to.

smnandre commented on src/TwigComponent/src/DependencyInjection/TwigComponentExtension.php:
```diff
@@ -217,6 +229,10 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->info('Enables the profiler for Twig Component (in debug mode)')
                     ->defaultValue('%kernel.debug%')
                 ->end()
+                ->booleanNode('short_syntax')
```
```suggestion
                ->booleanNode('short_tags')
```

More accurate than "syntax", which was a bit of a stretch (same thing to do in the otheer changes -- open to another naming idea)

smnandre commented on src/TwigComponent/tests/Unit/TwigPreLexerTest.php:
```diff
@@ -437,4 +452,79 @@ public static function getLexTests(): iterable
         ];
     }
+
+    public function getLexTestsWhithShortOptions()
```
```suggestion
    public static function getLexTestsWhithShortOptions()
```

smnandre commented:
I have real valid reason to not want this to be released right now. Sorry it has absolutely nothing to do with you @Halleck45 (or even the feature in itself).

But let's ignore them as the rest of the team decided. 

---

Focusing on the PR content.. good to me overall...  i'd like to make a final review when suggested changes are done (it will be easier to focus on things) 

Documentation/wording
- remove "JSX" mention (this syntax is in fact.... PascalCase HTML ^^)  
- replace "new syntax"  with one of the alternatives

Clean
- extract debug changes in another PR

And.... 
- add a lot more tests (using both <twig:Foo and <Foo in templates,  usage of ":",   comments,  attribute names, JSON content, etc etc)

Feel free if you need a hand on anything here 👍

Halleck45 committed:
Introduce a Short Tags system for TwigComponents

Halleck45 commented:
Hi! I believe I've addressed all the feedback.

I have:

- added quite a few tests. My imagination has limits, but I think I’ve covered a good range of cases
- removed the changes to the command
- standardized the naming to "ShortTags" and "shortTags"

I haven’t addressed the comments regarding autowiring. As mentioned in the comments, it’s currently not possible to use autowiring everywhere — the code frequently uses new, which makes it impossible without a major refactoring. In my opinion, rewriting the existing code is out of scope for this PR, which is why I prefer using an `addMethodCall`.
Moreover, many things depend on the environment, so pushing everything through autowiring doesn't always seem necessary for the existing code. In short, I don't think I’m in a position to handle this point in this branch.

Thanks !

nicolas-grekas commented:
My opinion on this PR:
I fear this is based on a very fragile technical base.

With this PR, the backend-side would be preempting every single tag that starts with an uppercase letter, effectively becoming incompatible with any frontend-side tag that'd use the same convention.

Since you mention JSX: it works with both custom and html5 tags *because* it has full knowledge of the DOM, since its parser is aligned to the HTML5 one.

But here, in twig, we're in confused-land from a parser PoV: twig knows nothing about HTML parsing rules, yet overlaps with some of them.

The `twig:` prefix made this somewhat acceptable: while it prevents writing anything on the frontend that'd use the same prefix, I think we can still consider us safe.

But any uppercase tag? This looks adventurous, to say the least.

A totally different template engine would be needed IMHO. Could be twig, with an HTML5 parser in front. But not just regexps.

I'm likely :-1: here, because of this. (I do understand the motivation, but building on sand wouldn't go very far)

Kocal commented:
I am closing due to inactivity, and for the reasons explained above. 
Thanks for the contribution!
